rfd/001-tcp-load-balancer-design-doc.md

```diff
@@ -0,0 +1,119 @@
+
+Generate private (client.crt) and public (client.key) keys for the client:
+$ openssl genrsa -out client.key
+$ openssl req -new -key client.key -out client_reqout.txt
+$ openssl x509 -req -in client_reqout.txt -days 3650 -sha256 -CAcreateserial -CA clentca.cert -CAkey clientcakey.pem -out client.crt		
+```
+
+For this project, a test client will be created and provided with a valid client cert for testing purposes. These initial dummy certificates and keys will be included (as plaintext) within this repository for the the benefit of reviewers, but should stored as secure secrets and managed outside of this repo in a future version.
+
+Additionally, the requirement of over-the-wire encryption is not mentioned, so even though the authentication requires certs and keys, the data sent after authentication may or may not be encrypted depending on what is most convenient at the point of implementation.
+
+### Rate Limiter
+This load balancer will implement a per-client connection rate limiter that tracks the number of client connections, and limits to n (configurable) active connections per client. It could track client identity based on attributes like IP address, and device identifier, but since this load balancer requires authentication (mTLS), we will identify unique clients based on a V5 UUID generated from the client's TLS certificate. 
+
+```
+clientID := uuid.NewV5(loadBalancerID, client.TLSCertificate().String())
+```
+
+Client connection counts can be stored in a `sync.Map` structure for quick modification, and safety across concurrent goroutines. Regarding this map choice, the [`sync` package documentation](https://pkg.go.dev/sync#Map) states the following: 
+> "The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex."
+
+Our use-case follows the #2 reason for leveraging `sync.Map` over a standard map with a Mutex; our disjoint sets of keys will be the unique client ids.
+
+
+### Upstream Authorization Management
+
+The load balancer will employ a simple authorization scheme which will deny access to specific upstreams for certain clients. This scheme will be statically defined in code. Once the load balancer has determined that the client is authenticated and within client connection limit (and prior to selecting the least connections host) it will filter out unauthorized hosts for that client. To simply demonstrate this capability, we store a map of `clientID`s (based on client certificate) to disallowed `hostID`s; because host configuration is outside the scope of this project, the order in which the host was registered with the load balancer will be used as the `hostID`.
```

Comment:
Yes, I'll get that static authorization scheme example added.

Here's what I'm thinking for a re-write of this section:

### Upstream Authorization Management

The load balancer will employ a simple authorization scheme which will deny access to specific upstreams for certain clients. This scheme will be statically defined in code, using something like the following `map` of `UnauthorizedClientHostRule`s to ensure an authorized host is selected. To simply demonstrate this capability, we store a map of `clientID:hostID` as an `UnauthorizedClientHostRule`. The only example rule which will be present for this project will be that client1 can never talk to host2. 

```

// UnauthorizedClientHostRule is a composite key made up of ClientID and HostID, and represents that the client is not authorized to connect to the host.
type UnauthorizedClientHostRule string

// FormatUnauthorizedClientHostRule builds a new unauthorizedClientHostRule from the given client and host IDs.
func BuildUnauthorizedClientHostRule(clientID uuid.UUID, hostID uuid.UUID) UnauthorizedClientHostRule {
	return UnauthorizedClientHostRule(fmt.Sprintf("%s:%s", clientID, hostID))
}

// unauthorized serves as a key-based lookup for unauthorized client-> host relationships.
// It will be checked every time a host is being selected. 
// If upstream hosts are to be registered dynamically, access to this map should be locked behind a mutex.
// For simplicity of this project, we'll only ever add rules during app startup, so the map is safe for concurrent reads.
var unauthorized = map[UnauthorizedClientHostRule]struct{}{}
```

Once the load balancer has determined that the client is authenticated and within client connection limit (and prior to selecting the least connections host) it will filter out unauthorized hosts for that client. See the following section for how the unauthorized map may be used for connection filtering.

...
(Then the following section will have the code example update checking the unauthorized map)

Reply:
> Are you stating that upstream hosts will also be static as well?

Yes, in order to limit scope of this project, the plan is to statically define hosts at startup, and the hostIDs will be configured at that time.

If a more dynamic approach is desired, there are many options that could be considered:
1. The Load Balancer could expose a separate port with an API to handle host registration.
2. A config file (updated by external process) could hold host information and either (a) be periodically inspected or (b) trigger reinspection when modified.
3. The Load Balancer could subscribe to an event queue with host updates, to which external processes could publish.
4. A cache or other data storage solution could be leveraged to persist host data, updated by some external processes.

As a customer, I would expect an easy solution for the load balancer to adapt to my backend deployment changes, without having to think about reconfiguring the load balancer (except for major server group changes, in the case of a shared load balancer). I think customer expectations could be met with any of the above options, depending on more specific requirements.

Are statically defined hosts satisfactory for this project?

Reply:
@rosstimothy to follow up on this conversation ☝️  ... Are statically defined hosts satisfactory for this project?

Reply:
I'm thinking of something like this:
```
// registerStaticUpstreamHosts adds n static hosts to the load balancer for testing and demonstration purposes.
func registerStaticUpstreamHosts(lb *server.LoadBalancer, n int) {
	for i := 0; i < n; i++ {
		u, err := upstream.New(dev.InitializeHelloHost(tcpNetwork, selectOpenPort).Addr().String())
		if err != nil {
			log.Fatal(err)
		}
		lb.AddHost(u)
	}
}

```
Which, I'll add to the section on demonstrating functionality at the bottom of this design doc.

For more details, the `dev.initializeHelloHost` func could look something like this:
```

// InitializeHelloHost is a temporary helper to simulate an upstream host that will print off any incoming data.
func InitializeHelloHost(tcpNetwork, address string) net.Listener {
	h, err := net.Listen(tcpNetwork, address)
	if err != nil {
		log.Fatalf("unable to listen on %s: %s", address, err)
	}

	go func() {
		for {
			conn, err := h.Accept()
			if err != nil {
				log.Printf("host was unable to accept incoming connection: %s", err)
				continue
			}

			PrintDataFromConnection(conn)
			conn.Close()
		}
	}()
	return h
}
```

Reply:
Yes, `selectOpenPort` would be `:0`, and good idea to throw in a grumpy host. I'll add a note about that to the design as well.

Reply:
I added a TODO on the example static host setup code so that we don't forget about this:
```
for {
      // TODO: Set up some interval, or predictable/controllable behavior where a host will not accept connection so that we can test the health checking process.
			conn, err := h.Accept()
```